python_modules/dagster/dagster/_core/events/__init__.py

```diff
@@ -886,10 +893,12 @@ def step_failure_event(
         step_failure_data: "StepFailureData",
         message=None,
     ) -> "DagsterEvent":
+        event_specific_data = step_failure_data._replace(asset_keys=step_context.asset_keys)
```

Comment:
this part feels a bit weird to me... that you'd have the executor construct the failure data but you'd be swapping it out by adding the info from the context.

Would you consider adding an optional parameter on the dagster event (e.g. step_asset_keys) rather than having it on the event specific data?

Reply:
That seems reasonable to me too, "asset keys" is a pretty top-level thing and it probably actually makes back-compat easier... it just seemed a little odd to add a new top-level field that only three events use, and is also already present in other events in the step specific data

Reply:
updating the callsites so you don't have to `_replace` i think is tractable too, theres not too many

Reply:
➕ 

reminding myself that we should make the storage interpretation of the presence/absence of these fields degrade nicely since it's user cloud version dependent